File: src/js/algorithms/merge-sort/merge-sort.js
function mergeSort(arr) {
  if (arr.length > 1) {
    const mid = Math.floor(arr.length / 2);
    let left = arr.slice(0, mid);
    let right = arr.slice(mid, arr.length);

    arr = merge(mergeSort(left), mergeSort(right));
  }

  return arr;
}

function merge(left, right) {
  let result = [];
  while (left.length && right.length) {
    let min = null;
    if (left[0] < right[0]) {
      min = left.shift();
    } else {
      min = right.shift();
    }
    result.push(min);
  }

  if (left.length) {
    result = result.concat(left);
  }

  if (right.length) {
    result = result.concat(right);
  }

  return result;
}

console.log(mergeSort([5, 4, 3, 2, 1]));

File: src/go/src/algorithms/merge-sort/merge-sort.go
package mergesort

import "fmt"

func mergeSort(s []int) []int {
	if len(s) == 1 {
		return s
	}

	mid := len(s) / 2

	left := s[:mid]
	right := s[mid:]
	new := make([]int, len(s))

	arr := merge(mergeSort(left), mergeSort(right), new)
	return arr
}

func merge(l []int, r []int, s []int) []int {
	i, j, k := 0, 0, 0

	for i < len(l) && j < len(r) {
		if l[i] < r[j] {
			s[k] = l[i]
			k++
			i++
		} else {
			s[k] = r[j]
			k++
			j++
		}
	}

	for i < len(l) {
		s[k] = l[i]
		k++
		i++
	}

	for j < len(r) {
		s[k] = r[j]
		k++
		j++
	}
	fmt.Println("print", s)
	return s
}

// Time complexity n (log n)
// Total levels = O(log(n))
// Work per level = O(n)
